### DISCUSSION TITLE
Add battery test sensor?

### DISCUSSION THREAD

**Discussion by jenshorn** [faanskit/ha-checkwatt#59] [Ideas]
As I believe you are already parsing the logs for the FCR-D state, would it be possible to also parse for the latest battery test run add a sensor with the result? Having that value, seeing it over time, and also give the possibility for automations/alerts in HA could be beneficial for multiple reasons.

**Comment by faanskit** [comment]
Thanks for the proposal, but it will be rejected.

You can easily create such a sensor using Templates as described in the [documentation](https://github.com/faanskit/ha-checkwatt?tab=readme-ov-file#use-templates). The FCR-D is available as an attribute on the CM10 sensor.

On top of that, FCR-D changes are already published as Events, again as described in the [documentation](https://github.com/faanskit/ha-checkwatt?tab=readme-ov-file#events).  Said events can be used to trigger automations. 

The following events are thrown and can be used to create automations: `fcrd_activated`, `fcrd_deactivated`, `fcrd_failed`
See example automation in the [documentation](https://github.com/faanskit/ha-checkwatt?tab=readme-ov-file#events).

**Comment by faanskit** [comment]
(Too quick to close...)
The issue will be converted to a discussion should feeback be needed.

**Comment by jenshorn** [comment]
I have seen the FCR-D events but I have failed to see any metrics from the latest battery test, ie 96.1% of possible power or however it is available in the logs. Are you saying that it's available somewhere already?

**Reply by faanskit** [reply]
When you upgrade to v0.2.0, which is currently available as beta [v0.2.0-RC.4) on Hacs, you will find a new sensor called CM10 Sensor. The CM10 Sensor can be `Offline`, `Under Test` or `Active`.
This sensor has the following FCR-D related attribute: `fcr_d_status`, `fcr_d_info`, `fcr_d_date`

The  `fcr_d_info` carries either the percentages from the last test, or the reason why it went back to `Under Test`.

This is how my CM10 sensor looks like now:
![image](https://github.com/faanskit/ha-checkwatt/assets/10439868/b2178412-8552-4089-b3f0-04848a6d2729)

And on top of that we have the events related to this:
```yaml
event_types:
  - fcrd_activated
  - fcrd_deactivated
  - fcrd_failed
event_type: fcrd_activated
attribution: Data provided by CheckWatt EnergyInBalance
device_class: fcrd
icon: mdi:battery-alert
friendly_name: NNN FCR-D State
```

What is new in v0.2.0 is that it also throws `fcrd_failed`, which in CM10 sensor attribute `fcrd_info` say something like "UP 51 Hz" or similar depending on which request was sent to the inverter, and failed.

**Reply by faanskit** [reply]
...before v2.0.0 the FCR-D status was available when you enabled **detailed attributes**, but was called `fcrd_percentage` on the Daily Monetary sensor.

**Reply by jenshorn** [reply]
Well that appears to be exactly what I'm looking for according to your dumps. I'm on v0.2.0-RC.4 and the CM10 sensor is not available by default. I guess I need to check the "Provide CM10 sensor" under "configure" for it to be available? There appears however to be some issue with the configure form as when I check that box I get `User input malformed: expected str for dictionary value @ data['cwr_name']` even though `Push data to CheckWattRank` is unchecked (I'm just making the assumption that those are related).

**Reply by jenshorn** [reply]
Also, pressing submit with no boxes checked gives the error: `User input malformed: expected bool for dictionary value @ data['cm10_sensor']`

**Reply by faanskit** [reply]
I actually don't know why it behaves like this, likely i corrupted something when I removed some config attributes.
The only way to fix this is to remove CheckWatt and re-install it. Good thing is that your data remains of you do that.

**Reply by faanskit** [reply]
#60 was created for this to make it transparent

**Reply by jenshorn** [reply]
Deleted and re-added the integration. `Provide CM10 sensor` is now selected by default and everything appears to work. Thank you for your time!

Btw, is there any documentation regarding each of the FCR-D Info `96,0/0,7/90,8 %`?

**Reply by faanskit** [reply]
I've seen it once referenced on facebook, but cannot find it.
I think it is FCR-D Up / Action Time / FCR-D Down

I can have mixed Up and Down.

There are supposed to be documentation available for partners for this.

**Reply by faanskit** [reply]
Found some info, do not have the reference
Discharge / Response time / Charge
Eg. 99,0/2,9/97,7

Targets (old?):
90% / 10 sec / 90%
